cmd/podman/images/scp.go

```diff
@@ -0,0 +1,350 @@
+	"github.com/docker/distribution/reference"
+	scpD "github.com/dtylman/scp"
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"github.com/spf13/cobra"
+	"golang.org/x/crypto/ssh"
+)
+
+var (
+	saveScpDescription = `Securely copy an image from one host to another.`
+	imageScpCommand    = &cobra.Command{
+		Use:               "scp [options] IMAGE [IMAGE...]",
+		Annotations:       map[string]string{registry.EngineMode: registry.ABIMode},
+		Long:              saveScpDescription,
+		Short:             "securely copy images",
+		RunE:              scp,
+		Args:              cobra.MinimumNArgs(1),
+		ValidArgsFunction: common.AutocompleteImages,
+		Example:           `podman image scp myimage:latest otherhost::`,
+	}
+)
+
+var (
+	scpOpts entities.ImageScpOptions
+)
+
+func init() {
+	registry.Commands = append(registry.Commands, registry.CliCommand{
+		Command: imageScpCommand,
+		Parent:  imageCmd,
+	})
+	scpFlags(imageScpCommand)
+}
+
+func scpFlags(cmd *cobra.Command) {
+	flags := cmd.Flags()
+	flags.BoolVarP(&scpOpts.Save.Quiet, "quiet", "q", false, "Suppress the output")
+}
+
+func scp(cmd *cobra.Command, args []string) (finalErr error) {
+	var (
+		//TODO add tag support for images
+		err error
+	)
+	if scpOpts.Save.Quiet { // set quiet for both load and save
+		scpOpts.Load.Quiet = true
+	}
+	f, err := ioutil.TempFile("", "podman") // open temp file for load/save output
+	if err != nil {
+		return err
+	}
+	defer os.Remove(f.Name())
+
+	scpOpts.Save.Output = f.Name()
+	scpOpts.Load.Input = scpOpts.Save.Output
+	if err := parse.ValidateFileName(saveOpts.Output); err != nil {
+		return err
+	}
+	confR, err := config.NewConfig("") // create a hand made config for the remote engine since we might use remote and native at once
+	if err != nil {
+		return errors.Wrapf(err, "could not make config")
+	}
+	abiEng, err := registry.NewImageEngine(cmd, args) // abi native engine
+	if err != nil {
+		return err
+	}
+
+	cfg, err := config.ReadCustomConfig() // get ready to set ssh destination if necessary
+	if err != nil {
+		return err
+	}
+	urls, iden, serv, err := parseArgs(args, cfg) // parses connection data and "which way" we are loading and saving
+	if err != nil {
+		return err
+	}
+	// TODO: Add podman remote support
+
+	uri := []*urlP.URL{}
+	for i, url := range urls {
+		urlt, err := urlP.Parse(url) // create an actual url to pass to exec command
+		if err != nil {
+			return err
+		}
+		uri = append(uri, urlt)
+		if uri[i].User.Username() == "" {
+			if uri[i].User, err = connection.GetUserInfo(uri[i]); err != nil {
+				return err
+			}
+		}
+	}
+
+	confR.Engine = config.EngineConfig{Remote: true, CgroupManager: "cgroupfs", ServiceDestinations: serv} // pass the service dest (either remote or something else) to engine
+
+	// TODO: add podman remote support
+
+	remoteFiles := []string{}
+	dials := []*ssh.Client{}
+	if scpOpts.ToRemote || scpOpts.FromRemote {
+		dials, remoteFiles, err = createConnections(uri, iden)
```

Comment:
Okay, I will move the dials down into the `toRemote` and `fromRemote` functions. I am still going to have to pass the uri/identity array entries though like I was doing before.

Reply:
oh one thing is though, the tempFiles does slightly matter since if we are doing `fromRemote` `toRemote` the file has to be kept track of.

Reply:
resolved nicely!!!

Reply:
I’m sorry, I should have been much more explicit about the data flow… I didn’t mean to imply that carrying state in the global `scpOpts` would be preferable.

Reply:
what do you mean by this? the most updated push deals with it nicely I believe. The multiple local uri and identity arrays make less sense since I re declared them 4 times.

Reply:
See https://github.com/containers/podman/pull/10828#discussion_r680062004 .